File: tests/test_map_hooks.js
// tests/test_map_hooks.js -- conformance for the Map.* primitive on the JS
// reference VM (picovm.js), driven through the English compiler (picoc.js).
// Run: node tests/test_map_hooks.js
const path = require('path');
const PicoCompile = require(path.join(__dirname, '..', 'vm', 'picoc.js'));
const PicoVM = require(path.join(__dirname, '..', 'vm', 'picovm.js'));

function runEnglish(src) {
  const r = PicoCompile.compileDebug(src, 'english');
  const vm = new PicoVM();
  vm.run(r.words.map((w) => w >>> 0));
  // Print writes each int as 4 big-endian bytes into vm.output.
  const out = vm.output, ints = [];
  for (let i = 0; i + 3 < out.length; i += 4) {
    ints.push(((out[i] << 24) | (out[i + 1] << 16) | (out[i + 2] << 8) | out[i + 3]) | 0);
  }
  return ints;
}

let pass = 0, fail = 0;
function eq(name, got, want) {
  const g = JSON.stringify(got), w = JSON.stringify(want);
  if (g === w) { pass++; console.log('PASS ' + name); }
  else { fail++; console.log('FAIL ' + name + ' -> got ' + g + ' want ' + w); }
}

// int keys + int values + count
eq('int keys/values + count', runEnglish(
  'Set h to Map.New().\n' +
  'Map.PutII(1, 100).\n' +
  'Map.PutII(2, 200).\n' +
  'Map.PutII(1, 111).\n' +          // overwrite keeps one entry
  'Print Map.GetII(1).\n' +
  'Print Map.GetII(2).\n' +
  'Print Map.Count().\n'
), [111, 200, 2]);

// has / del
eq('has + del', runEnglish(
  'Set h to Map.New().\n' +
  'Map.PutII(5, 50).\n' +
  'Print Map.HasI(5).\n' +
  'Print Map.HasI(9).\n' +
  'Map.DelI(5).\n' +
  'Print Map.HasI(5).\n' +
  'Print Map.Count().\n'
), [1, 0, 0, 0]);

// string keys + int values
eq('string keys', runEnglish(
  'Set h to Map.New().\n' +
  'Map.PutSI("qty", 42).\n' +
  'Map.PutSI("age", 7).\n' +
  'Print Map.GetSI("qty").\n' +
  'Print Map.GetSI("age").\n' +
  'Print Map.HasS("qty").\n' +
  'Print Map.HasS("nope").\n'
), [42, 7, 1, 0]);

// insertion-order enumeration
eq('enumeration order', runEnglish(
  'Set h to Map.New().\n' +
  'Map.PutSI("a", 10).\n' +
  'Map.PutSI("b", 20).\n' +
  'Map.PutSI("c", 30).\n' +
  'Print Map.ValAt(0).\n' +
  'Print Map.ValAt(1).\n' +
  'Print Map.ValAt(2).\n' +
  'Print Map.Count().\n'
), [10, 20, 30, 3]);

// null values distinguished from absent
eq('null vs absent', runEnglish(
  'Set h to Map.New().\n' +
  'Map.PutNullI(3).\n' +
  'Print Map.HasI(3).\n' +
  'Print Map.IsNullI(3).\n' +
  'Print Map.IsNullI(4).\n'
), [1, 1, 0]);

// two independent maps via Use
eq('two maps + Use', runEnglish(
  'Set a to Map.New().\n' +
  'Map.PutII(1, 10).\n' +
  'Set b to Map.New().\n' +
  'Map.PutII(1, 99).\n' +
  'Map.Use(a).\n' +
  'Print Map.GetII(1).\n' +
  'Map.Use(b).\n' +
  'Print Map.GetII(1).\n'
), [10, 99]);

// FNV-1a hash is stable/deterministic (same span hashes equal)
eq('hash determinism', runEnglish(
  'Set x to Map.Hash("Content-Type").\n' +
  'Set y to Map.Hash("Content-Type").\n' +
  'Print x minus y.\n'
), [0]);

console.log('\n' + (fail === 0 ? 'ALL ' + pass + ' PASSED' : fail + ' FAILED, ' + pass + ' passed'));
process.exit(fail === 0 ? 0 : 1);

File: tests/test_map_hooks.py
"""tests/test_map_hooks.py -- conformance for the Map.* primitive on the Python
reference VM (picoscript_vm.py), driven through the BASIC frontend. Mirrors the
JS suite (tests/test_map_hooks.js) so the two reference VMs agree.
"""
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from picoscript_basic import compile_basic          # noqa: E402
from picoscript_il import lower_to_bytecode_safe     # noqa: E402
from picoscript_vm import PicoVM                      # noqa: E402


def run_basic(src):
    words = lower_to_bytecode_safe(compile_basic(src))
    vm = PicoVM().run(words)
    out = []
    for c in vm.output:
        v = int.from_bytes(c, "big")
        out.append(v - 0x100000000 if v & 0x80000000 else v)
    return out


def test_int_keys_values_count():
    assert run_basic(
        "LET h = Map.New()\n"
        "Map.PutII(1, 100)\n"
        "Map.PutII(2, 200)\n"
        "Map.PutII(1, 111)\n"
        "PRINT Map.GetII(1)\n"
        "PRINT Map.GetII(2)\n"
        "PRINT Map.Count()\n"
    ) == [111, 200, 2]


def test_has_del():
    assert run_basic(
        "LET h = Map.New()\n"
        "Map.PutII(5, 50)\n"
        "PRINT Map.HasI(5)\n"
        "PRINT Map.HasI(9)\n"
        "Map.DelI(5)\n"
        "PRINT Map.HasI(5)\n"
        "PRINT Map.Count()\n"
    ) == [1, 0, 0, 0]


def test_string_keys():
    assert run_basic(
        "LET h = Map.New()\n"
        'Map.PutSI("qty", 42)\n'
        'Map.PutSI("age", 7)\n'
        'PRINT Map.GetSI("qty")\n'
        'PRINT Map.GetSI("age")\n'
        'PRINT Map.HasS("qty")\n'
        'PRINT Map.HasS("nope")\n'
    ) == [42, 7, 1, 0]


def test_enumeration_order():
    assert run_basic(
        "LET h = Map.New()\n"
        'Map.PutSI("a", 10)\n'
        'Map.PutSI("b", 20)\n'
        'Map.PutSI("c", 30)\n'
        "PRINT Map.ValAt(0)\n"
        "PRINT Map.ValAt(1)\n"
        "PRINT Map.ValAt(2)\n"
        "PRINT Map.Count()\n"
    ) == [10, 20, 30, 3]


def test_null_vs_absent():
    assert run_basic(
        "LET h = Map.New()\n"
        "Map.PutNullI(3)\n"
        "PRINT Map.HasI(3)\n"
        "PRINT Map.IsNullI(3)\n"
        "PRINT Map.IsNullI(4)\n"
    ) == [1, 1, 0]


def test_two_maps_use():
    assert run_basic(
        "LET a = Map.New()\n"
        "Map.PutII(1, 10)\n"
        "LET b = Map.New()\n"
        "Map.PutII(1, 99)\n"
        "Map.Use(a)\n"
        "PRINT Map.GetII(1)\n"
        "Map.Use(b)\n"
        "PRINT Map.GetII(1)\n"
    ) == [10, 99]


def test_hash_determinism():
    assert run_basic(
        'LET x = Map.Hash("Content-Type")\n'
        'LET y = Map.Hash("Content-Type")\n'
        "PRINT x - y\n"
    ) == [0]


if __name__ == "__main__":
    fails = 0
    for name, fn in sorted(globals().items()):
        if name.startswith("test_") and callable(fn):
            try:
                fn()
                print("PASS " + name)
            except AssertionError as e:
                fails += 1
                print("FAIL " + name + " -> " + str(e))
    print("\n" + ("ALL PASSED" if fails == 0 else f"{fails} FAILED"))
    sys.exit(1 if fails else 0)
